### ISSUE TITLE
Fix Async Issue

### ISSUE THREAD

**Issue by tOOlmaRR** [tOOlmaRR/toolmarrs-collectables#91]
**Description**: Address issues in current API with asynchronous requests

**Requirements**:

- Prerequisite: ["Article: Node.js Event Loop"](https://www.sitepoint.com/node-js-event-loop-guide/?utm_source=SitePoint&utm_medium=email&utm_campaign=SitePointWeekly) in Online Learning
- Determine if the issues are in my endpoints themselves, in the unit tests, or in both
- Currently, unit tests sometimes fail due to timeouts even though the requests themselves do not seem to take very long
- Address the issues as needed
- Original estimate: 4h; Current estimate: 4h

**Design:**

No solid plan was created.

**Test Plan:**

- Run the full test suite and ensure that all tests pass
- Build a list of URLs to test and ensure that they all return expected results and that they all return JSON without crashing
- Intentionally break the SQL connection (by setting the username, for instance to an incorrect value) and confirm that the tests do not timeout and that no handles are left open at the end of the test run

**Implementation Notes:**

- Tried a LOT of things to eliminate open handles
- Spent a LOT of time trying to stop running tests if a database connection fails (but may want to try again now that I’ve solved other issues)
- Updated a number of packages including jest (from v26 to v29 which now used the jest-circus test runner)
- Removed current version of node (10.19.0) and installed the latest (18.16.0) after installing nvm-windows
- Discovered I had numerous error handlers that did resulted in not returning a result, which accounted for the timeouts and the open handles I kept running across
- Discovered my SQL connection config had ‘stream = true’ which was causing query responses to not contain results.
- Converted all API method to explicitly return a promise rather than following the async/await pattern

**Release Notes:**

- Be sure to uninstall previous version of node, install nvm-windows and v18.16.0 of node.
- Be sure to reinstall packages via NPM

**Comment by tOOlmaRR** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Fix Async Issue](https://trello.com/c/oBgjHBAf/67-fix-async-issue)

**Comment by tOOlmaRR** [comment]
merged into master branch as part of v1.4 on 2025-02-15
